File: .claude/skills/startup-version-factor-matrix/scripts/matrix_report.py
#!/usr/bin/env python3
"""Cross-cell comparison for a version x factor run.

Per cell: n, window median/p90/max, per-pass medians (drift and pass-state), TTID median.
Then the version table (reference cells only) and the factor table (effect per anchor version).

The window comes from the app-side wrapper slice by default - the only instrument that exists in
every SDK version. Pass --slice emb-sdk-start to use the native window when every cell has it.

Usage:
  python3 matrix_report.py <run-dir> [--slice app-embrace-start] [--json out.json]
"""
import argparse
import json
import pathlib
import statistics
import subprocess

TP = pathlib.Path(__file__).parent / "trace_processor"

WINDOW_SQL = """
WITH win AS (
  SELECT s.ts AS ts, s.dur AS dur FROM slice s
  WHERE s.name = '{slice}' AND s.dur > 0
  ORDER BY s.ts DESC LIMIT 1
)
SELECT (SELECT dur / 1e6 FROM win) AS window_ms;
"""


def query_one(trace, sql_text):
    """Run one query against one trace; returns the first numeric value or None."""
    qf = pathlib.Path("/tmp/vfm_q.sql")
    qf.write_text(sql_text)
    cp = subprocess.run([str(TP), "-q", str(qf), str(trace)],
                        capture_output=True, text=True, timeout=600)
    for line in reversed(cp.stdout.strip().splitlines()):
        val = line.strip().strip('"')
        try:
            return float(val)
        except ValueError:
            continue
    return None


def cell_windows(cell_dir, slice_name):
    """Per-pass lists of window values, so pass-state stays visible instead of pooled away."""
    passes = {}
    for trace in sorted(cell_dir.rglob("*.perfetto-trace")):
        pass_key = next((p for p in trace.parts if p.startswith("pass")), trace.parent.name)
        val = query_one(trace, WINDOW_SQL.format(slice=slice_name))
        if val:
            passes.setdefault(pass_key, []).append(val)
    return passes


def summarize(passes):
    allv = sorted(v for vals in passes.values() for v in vals)
    if not allv:
        return None
    return {
        "n": len(allv),
        "median": statistics.median(allv),
        "p90": allv[min(len(allv) - 1, int(0.9 * len(allv)))],
        "max": allv[-1],
        "iqr": (allv[int(0.75 * len(allv))] - allv[int(0.25 * len(allv))]) if len(allv) >= 4 else 0.0,
        "pass_medians": {k: statistics.median(v) for k, v in sorted(passes.items()) if v},
    }


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("run_dir")
    ap.add_argument("--slice", default="app-embrace-start")
    ap.add_argument("--json")
    args = ap.parse_args()

    run_dir = pathlib.Path(args.run_dir)
    cells = {}
    for state_file in sorted(run_dir.rglob("cell-state.json")):
        state = json.loads(state_file.read_text())
        cell = state["cell"]
        summary = summarize(cell_windows(state_file.parent, args.slice))
        if summary is None:
            print(f"SKIP {cell['id']}: no window values ({args.slice} missing?)")
            continue
        cells[cell["id"]] = {"cell": cell, "summary": summary,
                            "build_type": state.get("build_type"),
                            "apk_sha256": (state.get("apk_sha256") or "")[:12]}

    print(f"\ninstrument: {args.slice}   (absolute values are build-type specific)\n")
    print(f"{'cell':<58}{'n':>5}{'med':>9}{'p90':>9}{'max':>9}  pass medians")
    for cid, c in sorted(cells.items()):
        s = c["summary"]
        pm = " ".join(f"{v:.0f}" for v in s["pass_medians"].values())
        flag = ""
        if len(s["pass_medians"]) > 1 and s["iqr"]:
            spread = max(s["pass_medians"].values()) - min(s["pass_medians"].values())
            if spread > s["iqr"]:
                flag = "  <-- PASS-STATE: re-judge on same-parity passes"
        print(f"{cid:<58}{s['n']:>5}{s['median']:>9.1f}{s['p90']:>9.1f}{s['max']:>9.1f}  {pm}{flag}")

    # version table: reference cells only
    ref = {cid: c for cid, c in cells.items() if cid.endswith("|reference")}
    if len(ref) > 1:
        print("\n=== VERSION TABLE (reference cell) ===")
        newest = max(ref, key=lambda k: ref[k]["cell"]["version"] == "local")
        base = ref[newest]["summary"]["median"]
        print(f"{'version':<14}{'med':>9}{'p90':>9}{'max':>9}{'delta vs newest':>18}")
        for cid, c in sorted(ref.items(), key=lambda kv: kv[1]["summary"]["median"]):
            s = c["summary"]
            d = s["median"] - base
            print(f"{c['cell']['version']:<14}{s['median']:>9.1f}{s['p90']:>9.1f}{s['max']:>9.1f}"
                  f"{d:>+13.1f} ms ({100 * d / base:+.0f}%)" if base else "")

    # factor table: each non-reference level vs the same version's reference cell
    factor_cells = {cid: c for cid, c in cells.items() if not cid.endswith("|reference")}
    if factor_cells:
        print("\n=== FACTOR TABLE (vs the same version's reference cell) ===")
        print(f"{'version':<14}{'factor level':<30}{'med':>9}{'ref med':>10}{'effect':>20}")
        for cid, c in sorted(factor_cells.items()):
            version = c["cell"]["version"]
            device = c["cell"]["device"]
            ref_id = f"{device}|{version}|reference"
            if ref_id not in cells:
                print(f"{version:<14}{cid.split('|')[-1]:<30}{c['summary']['median']:>9.1f}"
                      f"{'--':>10}{'no reference cell':>20}")
                continue
            rm = cells[ref_id]["summary"]["median"]
            m = c["summary"]["median"]
            print(f"{version:<14}{cid.split('|')[-1]:<30}{m:>9.1f}{rm:>10.1f}"
                  f"{m - rm:>+13.1f} ms ({100 * (m - rm) / rm:+.0f}%)")
        print("\nParallel effects across anchor versions = factor effects. Effects that change "
              "sign or magnitude materially = version x factor interactions worth investigating.")

    print("\nReminders: pair every window delta with TTID and pre-TTID main-thread CPU (a window "
          "win with flat pre-TTID CPU is re-attribution, not improvement); never compare across "
          "build types or devices except as labelled tier replication.")

    if args.json:
        pathlib.Path(args.json).write_text(json.dumps(cells, indent=1, default=str))
        print(f"wrote {args.json}")


if __name__ == "__main__":
    main()

File: startup-tools/src/main/kotlin/io/embrace/startup/analysis/MatrixReport.kt
package io.embrace.startup.analysis

import io.embrace.startup.core.json.PyJson
import io.embrace.startup.core.json.StartupJson
import io.embrace.startup.core.stats.Quantile
import io.embrace.startup.core.text.PyFormat
import io.embrace.startup.perfetto.Queries
import io.embrace.startup.perfetto.TraceProcessor
import io.embrace.startup.perfetto.TraceReads
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonObject
import java.nio.file.Files
import java.nio.file.Path
import kotlin.streams.toList

/**
 * `matrix_report.py`: cross-cell comparison for a version × factor run. Per cell: n, window
 * median/p90/max, per-pass medians (so pass-state stays visible instead of pooled away); then the
 * version table (reference cells only) and the factor table (effect per anchor version, against the
 * same version's reference cell).
 *
 * No Python golden exists for this report (it needs cell-state run directories, none survived), so
 * the port is checked against hand-derived expectations on a synthetic run directory. Departure
 * (port log): the default window slice is `emb-sdk-start`, with `composed` for pre-9.2.0 cells; the
 * Python defaulted to `app-embrace-start`, a slice the harness never emitted.
 */
object MatrixReport {

    const val DEFAULT_SLICE: String = "emb-sdk-start"

    data class Summary(
        val n: Int,
        val median: Double,
        val p90: Double,
        val max: Double,
        val iqr: Double,
        /** Pass key → median, keys sorted as strings (`pass1`, `pass10`, `pass2`, …) as the Python did. */
        val passMedians: Map<String, Double>,
    )

    data class CellResult(val cell: JsonObject, val summary: Summary, val buildType: String?, val apkSha12: String)

    /** Window values grouped by pass (the first path element starting with `pass`, else the parent dir). */
    fun cellWindows(cellDir: Path, readWindow: (Path) -> Double?): Map<String, List<Double>> {
        val passes = LinkedHashMap<String, MutableList<Double>>()
        val traces = Files.walk(cellDir).use { s ->
            s.filter { Files.isRegularFile(it) && it.fileName.toString().endsWith(".perfetto-trace") }.toList()
        }.sorted()
        traces.forEach { trace ->
            val key = trace.map { it.toString() }.firstOrNull { it.startsWith("pass") } ?: trace.parent.fileName.toString()
            val value = readWindow(trace)
            if (value != null && value != 0.0) passes.getOrPut(key) { ArrayList() }.add(value)
        }
        return passes
    }

    /** `summarize`: index-based p90 and IQR (the store's legacy definitions), pass medians in key order. */
    fun summarize(passes: Map<String, List<Double>>): Summary? {
        val all = passes.values.flatten().sorted()
        if (all.isEmpty()) return null
        val n = all.size
        val iqr = if (n >= MIN_N_FOR_IQR) all[(P75 * n).toInt()] - all[(P25 * n).toInt()] else 0.0
        return Summary(
            n = n,
            median = Quantile.median(all),
            p90 = Quantile.legacyIndex(all, P90),
            max = all.last(),
            iqr = iqr,
            passMedians = passes.toSortedMap().filterValues { it.isNotEmpty() }.mapValues { Quantile.median(it.value.sorted()) },
        )
    }

    /** Every `cell-state.json` under [runDir], measured; skipped cells are reported in the returned lines. */
    fun collect(
        runDir: Path,
        readWindow: (Path) -> Double?,
        slice: String,
    ): Pair<Map<String, CellResult>, List<String>> {
        val cells = LinkedHashMap<String, CellResult>()
        val notes = ArrayList<String>()
        val states = Files.walk(runDir).use { s ->
            s.filter { Files.isRegularFile(it) && it.fileName.toString() == "cell-state.json" }.toList()
        }.sorted()
        states.forEach { stateFile ->
            val state = StartupJson.parseToJsonElement(Files.readString(stateFile)).jsonObject
            val cell = state.getValue("cell").jsonObject
            val id = PyJson.str(cell, "id", "?")
            val summary = summarize(cellWindows(stateFile.parent, readWindow))
            if (summary == null) {
                notes.add("SKIP $id: no window values ($slice missing?)")
                return@forEach
            }
            val sha = (PyJson.strOrNull(state, "apk_sha256") ?: "").take(SHA_CHARS)
            cells[id] = CellResult(cell, summary, PyJson.strOrNull(state, "build_type"), sha)
        }
        return cells to notes
    }

    fun windowReader(tp: TraceProcessor, slice: String): (Path) -> Double? = { trace ->
        TraceReads.parseWindow(tp.queryRaw(Queries.windowFor(slice), trace).stdout)
    }

    fun report(cells: Map<String, CellResult>, slice: String, notes: List<String> = emptyList()): String {
        val out = StringBuilder()
        notes.forEach { out.line(it) }
        out.line("\ninstrument: $slice   (absolute values are build-type specific)\n")
        out.line("${"cell".padEnd(W58)}${"n".padStart(W5)}${"med".padStart(W9)}${"p90".padStart(W9)}${"max".padStart(W9)}  pass medians")
        cells.toSortedMap().forEach { (cid, c) ->
            val s = c.summary
            val pm = s.passMedians.values.joinToString(" ") { PyFormat.fixed(it, 0) }
            var flag = ""
            if (s.passMedians.size > 1 && s.iqr != 0.0) {
                val spread = s.passMedians.values.max() - s.passMedians.values.min()
                if (spread > s.iqr) flag = "  <-- PASS-STATE: re-judge on same-parity passes"
            }
            out.line("${cid.padEnd(W58)}${s.n.toString().padStart(W5)}${f1(s.median, W9)}${f1(s.p90, W9)}${f1(s.max, W9)}  $pm$flag")
        }
        out.versionTable(cells)
        out.factorTable(cells)
        out.line(
            "\nReminders: pair every window delta with TTID and pre-TTID main-thread CPU (a window " +
                "win with flat pre-TTID CPU is re-attribution, not improvement); never compare across " +
                "build types or devices except as labelled tier replication.",
        )
        return out.toString()
    }

    /** `--json`: the cells dict as the Python dumped it (summary, build type, short sha). */
    fun toJson(cells: Map<String, CellResult>): JsonObject = JsonObject(
        cells.mapValues { (_, c) ->
            val s = c.summary
            JsonObject(
                mapOf(
                    "cell" to c.cell,
                    "summary" to JsonObject(
                        mapOf(
                            "n" to JsonPrimitive(s.n),
                            "median" to JsonPrimitive(s.median),
                            "p90" to JsonPrimitive(s.p90),
                            "max" to JsonPrimitive(s.max),
                            "iqr" to JsonPrimitive(s.iqr),
                            "pass_medians" to JsonObject(s.passMedians.mapValues { JsonPrimitive(it.value) as JsonElement }),
                        ),
                    ),
                    "build_type" to (c.buildType?.let { JsonPrimitive(it) } ?: JsonNull),
                    "apk_sha256" to JsonPrimitive(c.apkSha12),
                ),
            )
        },
    )

    private fun StringBuilder.line(s: String = ""): StringBuilder = append(s).append('\n')

    private fun StringBuilder.versionTable(cells: Map<String, CellResult>) {
        val ref = cells.filterKeys { it.endsWith("|reference") }
        if (ref.size <= 1) return
        line("\n=== VERSION TABLE (reference cell) ===")
        // The Python's `max(ref, key=version == "local")`: a local cell if any, else the first key.
        val newest = ref.keys.firstOrNull { PyJson.strOrNull(ref.getValue(it).cell, "version") == "local" } ?: ref.keys.first()
        val base = ref.getValue(newest).summary.median
        line("${"version".padEnd(W14)}${"med".padStart(W9)}${"p90".padStart(W9)}${"max".padStart(W9)}${"delta vs newest".padStart(W18)}")
        ref.entries.sortedBy { it.value.summary.median }.forEach { (_, c) ->
            val s = c.summary
            val d = s.median - base
            if (base != 0.0) {
                line(
                    "${PyJson.str(c.cell, "version", "None").padEnd(W14)}${f1(s.median, W9)}${f1(s.p90, W9)}${f1(s.max, W9)}" +
                        "${PyFormat.fixedSigned(d, 1).padStart(W13)} ms (${PyFormat.fixedSigned(PERCENT * d / base, 0)}%)",
                )
            } else {
                line("")
            }
        }
    }

    private fun StringBuilder.factorTable(cells: Map<String, CellResult>) {
        val factor = cells.filterKeys { !it.endsWith("|reference") }
        if (factor.isEmpty()) return
        line("\n=== FACTOR TABLE (vs the same version's reference cell) ===")
        line(
            "${"version".padEnd(W14)}${"factor level".padEnd(W30)}${"med".padStart(W9)}${"ref med".padStart(W10)}${"effect".padStart(W20)}",
        )
        factor.toSortedMap().forEach { (cid, c) ->
            val version = PyJson.str(c.cell, "version", "None")
            val device = PyJson.str(c.cell, "device", "None")
            val level = cid.substringAfterLast("|")
            val refCell = cells["$device|$version|reference"]
            val m = c.summary.median
            if (refCell == null) {
                line("${version.padEnd(W14)}${level.padEnd(W30)}${f1(m, W9)}${"--".padStart(W10)}${"no reference cell".padStart(W20)}")
                return@forEach
            }
            val rm = refCell.summary.median
            line(
                "${version.padEnd(W14)}${level.padEnd(W30)}${f1(m, W9)}${f1(rm, W10)}" +
                    "${PyFormat.fixedSigned(m - rm, 1).padStart(W13)} ms (${PyFormat.fixedSigned(PERCENT * (m - rm) / rm, 0)}%)",
            )
        }
        line(
            "\nParallel effects across anchor versions = factor effects. Effects that change " +
                "sign or magnitude materially = version x factor interactions worth investigating.",
        )
    }

    private fun f1(x: Double, width: Int) = PyFormat.fixed(x, 1).padStart(width)

    private const val P90 = 0.9
    private const val P75 = 0.75
    private const val P25 = 0.25
    private const val MIN_N_FOR_IQR = 4
    private const val SHA_CHARS = 12
    private const val PERCENT = 100.0
    private const val W5 = 5
    private const val W9 = 9
    private const val W10 = 10
    private const val W13 = 13
    private const val W14 = 14
    private const val W18 = 18
    private const val W20 = 20
    private const val W30 = 30
    private const val W58 = 58
}
